File: jc_timber_mobile/lib/models/marketplace_listing.dart
/// Marketplace listing - matches backend MarketplaceListing schema.
class MarketplaceListing {
  final String id;
  final String title;
  final double price;
  final String category;
  final String condition;
  final String description;
  final String location;
  final LocationCoords? locationCoords;
  final String? imageUrl;
  final String status;
  final String? userId;
  final String? userName;
  final String? userEmail;
  final DateTime? createdAt;

  const MarketplaceListing({
    required this.id,
    required this.title,
    required this.price,
    required this.category,
    required this.condition,
    required this.description,
    required this.location,
    this.locationCoords,
    this.imageUrl,
    required this.status,
    this.userId,
    this.userName,
    this.userEmail,
    this.createdAt,
  });

  factory MarketplaceListing.fromJson(Map<String, dynamic> json) {
    final user = json['user'];
    String? userId;
    String? userName;
    String? userEmail;
    if (user is Map<String, dynamic>) {
      userId = user['_id']?.toString() ?? user['id']?.toString();
      userName = user['name'] as String?;
      userEmail = user['email'] as String?;
    }

    LocationCoords? coords;
    final lc = json['locationCoords'];
    if (lc is Map<String, dynamic>) {
      final lat = (lc['lat'] as num?)?.toDouble();
      final lon = (lc['lon'] as num?)?.toDouble();
      if (lat != null && lon != null) {
        coords = LocationCoords(lat: lat, lon: lon);
      }
    }

    return MarketplaceListing(
      id: (json['_id'] ?? json['id']).toString(),
      title: json['title'] as String? ?? '',
      price: (json['price'] as num?)?.toDouble() ?? 0,
      category: json['category'] as String? ?? '',
      condition: json['condition'] as String? ?? '',
      description: json['description'] as String? ?? '',
      location: json['location'] as String? ?? '',
      locationCoords: coords,
      imageUrl: json['imageUrl'] as String?,
      status: json['status'] as String? ?? 'active',
      userId: userId,
      userName: userName,
      userEmail: userEmail,
      createdAt: json['createdAt'] != null
          ? DateTime.tryParse(json['createdAt'].toString())
          : null,
    );
  }

  static const conditionLabels = {
    'new': 'New',
    'used-like-new': 'Used - Like New',
    'used-good': 'Used - Good',
    'fair': 'Fair',
  };

  String get conditionLabel =>
      conditionLabels[condition] ?? condition;
}

class LocationCoords {
  final double lat;
  final double lon;

  const LocationCoords({required this.lat, required this.lon});
}

File: server/src/models/MarketplaceListing.js
import mongoose from "mongoose";

const marketplaceListingSchema = new mongoose.Schema({
  user: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  title: {
    type: String,
    required: true,
    trim: true
  },
  price: {
    type: Number,
    required: true,
    min: 0
  },
  category: {
    type: String,
    required: true,
    trim: true
  },
  condition: {
    type: String,
    required: true,
    enum: ['new', 'used-like-new', 'used-good', 'fair']
  },
  description: {
    type: String,
    required: true,
    trim: true
  },
  location: {
    type: String,
    required: true,
    trim: true
  },
  locationCoords: {
    lat: {
      type: Number,
      required: true
    },
    lon: {
      type: Number,
      required: true
    }
  },
  image: {
    data: { type: String },
    url: { type: String },
    publicId: { type: String },
    contentType: {
      type: String,
      required: true
    },
    filename: {
      type: String,
      required: true
    }
  },
  status: {
    type: String,
    enum: ['pending', 'approved', 'rejected', 'active', 'sold', 'inactive'],
    default: 'pending'
  },
  paymentStatus: {
    type: String,
    enum: ['pending', 'paid', 'failed', 'refunded'],
    default: 'pending'
  },
  payment: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'MarketplacePayment',
    default: null
  },
  paymentSlotIndex: {
    type: Number,
    default: null // Index of this listing in the payment (1, 2, or 3)
  },
  approvedAt: {
    type: Date,
    default: null
  },
  approvedBy: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    default: null
  },
  rejectionReason: {
    type: String,
    default: null
  }
}, {
  timestamps: true
});

marketplaceListingSchema.pre("save", function (next) {
  if (this.image && !this.image.data && !this.image.url) {
    return next(new Error("Listing image must have either data or url"));
  }
  next();
});

// Index for better query performance
marketplaceListingSchema.index({ user: 1, status: 1 });
marketplaceListingSchema.index({ status: 1, createdAt: -1 });
marketplaceListingSchema.index({ category: 1, status: 1 });

export default mongoose.model("MarketplaceListing", marketplaceListingSchema);
